library/alloc/src/vec/source_iter_marker.rs

```diff
@@ -31,6 +34,34 @@ where
             return SpecFromIterNested::from_iter(iterator);
         }
 
+        Self::from_iter_marker_impl::<I>(iterator)
+    }
+
+    default fn try_from_iter(iterator: I) -> Result<Self, TryReserveError> {
+        // Additional requirements which cannot expressed via trait bounds. We rely on const eval
+        // instead:
+        // a) no ZSTs as there would be no allocation to reuse and pointer arithmetic would panic
+        // b) size match as required by Alloc contract
+        // c) alignments match as required by Alloc contract
+        if mem::size_of::<T>() == 0
+            || mem::size_of::<T>()
+                != mem::size_of::<<<I as SourceIter>::Source as AsIntoIter>::Item>()
+            || mem::align_of::<T>()
+                != mem::align_of::<<<I as SourceIter>::Source as AsIntoIter>::Item>()
+        {
+            // fallback to more generic implementations
+            return SpecFromIterNested::try_from_iter(iterator);
+        }
+
+        Ok(Self::from_iter_marker_impl::<I>(iterator))
+    }
```

Comment:
Can you dedup this too? I don't think this trait is public, which could help.

Reply:
I've gone back and deduped all of these: it clean up a ton code.